added image_generator

JoashMadapogu opened the pull request:
## Description

Provide a concise summary of the changes in this PR. Include the motivation behind these changes and any relevant context. 

## Related Issue

If this PR addresses an issue, link to it here.

## Type of Change

Please select the type(s) of change that apply and delete those that do not. 

- [ ] **Bug fix**: A non-breaking change that fixes an issue.
- [x] **New feature**: A non-breaking change that adds functionality.
- [ ] **Breaking change**: A change that causes existing functionality to not work as expected.
- [ ] **Documentation update**: Changes or updates to documentation.
- [ ] **Code style update**: Changes that do not affect the meaning of the code (e.g., formatting).
- [ ] **Refactoring**: A code change that neither fixes a bug nor adds a feature.
- [ ] **Performance improvement**: A change that improves performance.
- [ ] **Test enhancement**: Adding or updating tests; no production code change.
- [ ] **Chore**: Changes to the build process or auxiliary tools; no production code change.
- [ ] **Other**: (please describe)

## Proposed Solution

Describe your code changes in detail. Explain how you implemented your solution and any design decisions you made. 

## How to Test

 Provide instructions on how to test these changes. Include details on test configurations, test cases, and expected outcomes. 

## Unit Tests

 List the unit tests added or modified to verify your changes. 

## Documentation Updates

Indicate whether documentation needs to be updated due to this PR. 

- [ ] Yes
- [ ] No

If yes, describe what documentation updates are needed and link to the relevant documentation.

## Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

## Additional Information

 Add any other information that might be useful for the reviewers.

Commit:
added image_generator

JoashMadapogu commented:
New “image-generator” tool entry
• Added to tools_config.json under the key "image-generator", pointing to tools.image_generator.core and its metadata.json.

Core logic (tools.py)
• Prompt handling:
– read_prompt_file() loads prompt templates from your prompt/ directory.
– check_prompt_safety() calls Gemini via LangChain-Google to flag unsafe or kid-related prompts.
– executor_image_generator() ties it all together: runs the safety check, uses Gemini to enhance the prompt, calls Vertex AI’s Imagen to generate an image (with adult-only filtering), and raises a clear error if no image is returned.

• Cloud storage:
– save_to_cloud_storage() uses your FirebaseManager to upload the generated image bytes and return a signed URL.

Wrapper entry point (core.py)
• Validates inputs with a Pydantic ImageGenerationInput (enforcing non-empty base_prompt, grade_level, subject).
• Pulls API key and project ID from environment (GOOGLE_API_KEY, PROJECT_ID).
• Calls executor_image_generator() and then validates its output with ImageGenerationOutput before returning it.

Schemas (schemas.py)
• Defines ImageGenerationInput and ImageGenerationOutput models.
• Shared non-empty string validator on required fields, so empty strings or None get a clear “cannot be empty” error.

Tests
• test_tools.py covers unit-testing of each helper function and error path in tools.py (prompt-reading, safety check, language mapping, Firebase upload, and generator failures).
• test_core.py mocks environment variables and the executor_image_generator() call to verify full-pipeline behavior, input validation errors, missing-env errors, and output validation.

Firebase support (firebase.py)
• Singleton FirebaseManager initializes the Admin SDK from the service-account JSON, uploads image bytes to the configured bucket, and returns a signed URL valid for seven days (this can be changed).

stevenrayhinojosa-gmail-com commented:
Hi, thank you for the fantastic contribution. I will run this locally in the mean while at your earliest convenience please add more information to the following sections as these sections help us keep track of changes:
Description
Related Issue
Type of Change
Proposed Solution
How to Test
Unit Tests
Documentation Updates
Additional Information

Again, thank you for your hard work!

JoashMadapogu commented:
Description
Introduces a new “image-generator” tool into our framework. This includes:
A core implementation (tools.image_generator.core) that checks prompt safety via Gemini, enhances prompts, invokes Vertex AI’s Imagen model, and uploads results to Firebase
A Pydantic-backed wrapper (app.tools.image_generator.core) validating inputs/outputs.
Supporting schemas (schemas.py) with non-empty validators.
A singleton FirebaseManager for signed-URL uploads.
Updated tools_config.json to register the new tool.
Comprehensive unit tests covering prompt loading, safety checks, language mapping, image generation pipeline, environment-variable guards, and error paths.

Type of Change
✨ New feature
🧪 Testing (added new unit tests)
Proposed Solution
Core Logic: Implemented executor_image_generator() to orchestrate prompt safety, prompt enhancement, image generation, and upload.
Wrapper: Created executor() entry point with Pydantic validation, environment-variable checks, and error wrapping.
Schemas: Added ImageGenerationInput/ImageGenerationOutput with field validators to ensure required inputs are non-empty.
Configuration: Registered the tool in tools_config.json with its metadata.json.
Firebase Support: Built FirebaseManager for authenticated uploads and signed-URL generation.
Testing: Wrote test_tools.py and test_core.py to cover success and failure scenarios across the pipeline.

How to Test
Set environment variables:
export GOOGLE_API_KEY=<your-key>  
export PROJECT_ID=<your-gcp-project>  
export GOOGLE_APPLICATION_CREDENTIALS=<path-to-service-account.json>  
export FIREBASE_STORAGE_BUCKET=<your-firebase-bucket>  
Run all tests:
Pytest

Unit Tests
test_tools.py:
read_prompt_file() success/failure
check_prompt_safety() safe/unsafe responses
map_language() mappings
executor_image_generator() happy path, unsafe prompt, no images, and upload failures
test_core.py:
Full-pipeline success with mocks
Missing-env errors for GOOGLE_API_KEY and PROJECT_ID
Input validation errors for empty or null fields
Proper error wrapping for ToolExecutorError and unexpected exceptions

Documentation Updates
Added metadata.json for the image-generator UI schema.
Updated tools_config.json with the new tool entry.
(Coming in a follow-up) I’ll add a brief section to the README describing usage and environment setup.
Additional Information
The Firebase signed URL defaults to 7-day expiration but can be configured via expiration_minutes.
Error messages are designed to help end-users rephrase unsafe prompts or provide missing API configuration.
Future improvements could include support for multiple aspect ratios or batch image requests.

stevenrayhinojosa-gmail-com commented:
Hi, Thanks i ran a Pytest and the image generator passed. With these PR's we like to work on the aesthetics of PR's like documentation etc. thank you for the contribution!
